db/compaction/compaction_picker.cc

```diff
@@ -39,32 +39,46 @@ bool FindIntraL0Compaction(const std::vector<FileMetaData*>& level_files,
                            size_t min_files_to_compact,
                            uint64_t max_compact_bytes_per_del_file,
                            uint64_t max_compaction_bytes,
-                           CompactionInputFiles* comp_inputs) {
-  size_t compact_bytes = static_cast<size_t>(level_files[0]->fd.file_size);
-  uint64_t compensated_compact_bytes = level_files[0]->compensated_file_size;
+                           CompactionInputFiles* comp_inputs,
+                           SequenceNumber earliest_mem_seqno) {
+
+  // Do not pick ingested file when there is at least one memtable not flushed which of seqno is overlap with the sst.
+  size_t start = 0;
+  while (start < level_files.size() && level_files[start]->being_compacted &&
```

Comment:
I think there should be `being_compacted ||`. I think it might be clearer to perform the conditionals one at a time in the loop:

```
for (; start < level_files_.size(); ++start) {
  if (level_files[start]->being_compacted) {
    continue;
  }
  if (level_files[start]->fd.largest_seqno >= earliest_mem_seqno) {
    continue;
  }
  break;
}
```

Reply:
If anything in the prefix that is being skipped due to earliest_mem_seqno is already being_compacted, I think we can return false. And I am not sure if that can even happen:
- if it is an intra-L0 compaction then earliest_mem_seqno must have not been monotonic
- if it is an L0 to a lower level compaction we must have skipped over the non-flushed memtable to include a later sstable which seems incorrect.

Reply:
I think `return false` when we reach a `start` that is undergoing compaction is good. It is aligned with what we are doing in `LevelCompactionBuilder::PickIntraL0Compaction()` when the file at index zero is undergoing compaction. Perhaps the latter can be removed once we have the check here.